app/forms/security_event_form.rb

```diff
@@ -0,0 +1,201 @@
+# rubocop:disable Metrics/ClassLength
+# Handles SET events (Security Event Tokens)
+class SecurityEventForm
+  include Rails.application.routes.url_helpers
+  include ActiveModel::Model
+
+  # From https://tools.ietf.org/html/draft-ietf-secevent-http-push-00#section-2.3
+  module ErrorCodes
+    JWS = 'jws'.freeze
+    JWT_AUD = 'jwtAud'.freeze
+    JWT_CRYPTO = 'jwtCrypto'.freeze
+    JWT_HDR = 'jwtHdr'.freeze
+    JWT_PARSE = 'jwtParse'.freeze
+    SET_DATA = 'setData'.freeze
+    SET_TYPE = 'setType'.freeze
+  end
+
+  validate :validate_iss
+  validate :validate_aud
+  validate :validate_event_type
+  validate :validate_subject_type
+  validate :validate_sub
+  validate :validate_typ
+  validate :validate_exp
+  validate :validate_jwt
+
+  def initialize(body:)
+    @body = body
+    @jwt_payload, @jwt_headers = parse_jwt
+  end
+
+  def submit
+    success = valid?
+
+    if success
+      SecurityEvent.create(
+        user: user,
+        event_type: event_type,
+        jti: jti,
+        issuer: service_provider.issuer,
+      )
+    end
+
+    FormResponse.new(success: success, errors: errors.messages, extra: extra_analytics_attributes)
+  end
+
+  def error_code
+    return if valid?
+
+    @error_code ||= ErrorCodes::SET_DATA
+  end
+
+  def description
+    return if valid?
+
+    errors.full_messages.join(', ')
+  end
+
+  private
+
+  attr_reader :body, :jwt_payload, :jwt_headers
+
+  # @return [Array(Hash, Hash)] parses JWT into [payload, headers]
+  def parse_jwt
+    JWT.decode(body, nil, false, algorithm: 'RS256', leeway: Float::INFINITY)
+  rescue JWT::DecodeError
+    @error_code = ErrorCodes::JWT_PARSE
```

Comment:
I think is the only place an `@error_code` gets set outside of the `validate_` process. Would it be possible to have `check_jwt_parse_error` look at `@jwt_payload, @jwt_header` during validation and set the error code there instead of in the constructor?

Reply:
Yeah, I didn't feel great about putting this here, either

My issue is that inside of `check_jwt_parse_error`, we could check for `@jwt_payload == {}` but it would be impossible to tell the difference between an empty payload and one that failed to parse. I guess I could set `@jwt_payload = nil` but then we have to add nil-safe access or early bails everywhere else in the class?

How strongly do you think we should change this?

Reply:
Ah, right I hadn't considered that. The other solution that came to mind would be moving `validate_jwt` to the top of the validators so it goes first and doing the parse/assign there instead of the constructor, but the ordering having a significant effect on behavior feels weird too.

I do think this is preferable to adding more nil checks.